src/nupic/research/temporal_memory.py

```diff
@@ -123,416 +214,197 @@ def compute(self, activeColumns, learn=True):
-                           winnerCells,
-                           prevWinnerCells,
-                           self.connections,
-                           predictedInactiveCells,
-                           prevMatchingSegments)
+    """
+    prevActiveCells = sorted(self.activeCells)
+    prevWinnerCells = sorted(self.winnerCells)
+
+    activeColumns = sorted(activeColumns)
+
+    self.activeCells = set()
+    self.winnerCells = set()
+
+    for excitedColumn in excitedColumnsGenerator(activeColumns,
+                                                 self.activeSegments,
+                                                 self.matchingSegments,
+                                                 self.cellsPerColumn,
+                                                 self.connections):
+      if excitedColumn["isActiveColumn"]:
+        if (excitedColumn["activeSegmentsBegin"] !=
+            excitedColumn["activeSegmentsEnd"]):
+
+          cellsToAdd = self.activatePredictedColumn(excitedColumn, learn,
+                                                    prevActiveCells)
+          self.activeCells.update(cellsToAdd)
+          self.winnerCells.update(cellsToAdd)
+        else:
+          (cellsToAdd,
+           winnerCell) = self.burstColumn(excitedColumn, learn,
+                                          prevActiveCells, prevWinnerCells)
+          self.activeCells.update(cellsToAdd)
+          self.winnerCells.add(winnerCell)
+      else:
+        if learn:
+          self.punishPredictedColumn(excitedColumn, prevActiveCells)
 
     (activeSegments,
-     predictiveCells,
-     matchingSegments,
-     matchingCells) = self.computePredictiveCells(activeCells, self.connections)
+     matchingSegments) = self.connections.computeActivity(self.activeCells,
+                                                          self.connectedPermanence,
+                                                          self.activationThreshold,
+                                                          0.0,
+                                                          self.minThreshold)
 
-    self.activeCells = activeCells
-    self.winnerCells = winnerCells
     self.activeSegments = activeSegments
-    self.predictiveCells = predictiveCells
     self.matchingSegments = matchingSegments
-    self.matchingCells = matchingCells
-
 
   def reset(self):
     """
     Indicates the start of a new sequence. Resets sequence state of the TM.
     """
     self.activeCells = set()
-    self.predictiveCells = set()
     self.activeSegments = set()
     self.winnerCells = set()
 
 
-  # ==============================
-  # Phases
-  # ==============================
-
-  def activateCorrectlyPredictiveCells(self,
-                                       prevPredictiveCells,
-                                       prevMatchingCells,
-                                       activeColumns):
+  def activatePredictedColumn(self, excitedColumn, learn, prevActiveCells):
```

Comment:
Here's a conversation topic: maybe these should be functions with lots of parameters, rather than methods that access the `self`.

That's how I did it in C++, not because it's necessary, but because it makes it easy to reason about. For example, it's really useful being able to see from the `compute` method where the random number generator gets consumed: https://github.com/numenta/nupic.core/blob/a83ed19fc2dcf901352da865d0b9f42892b8fa60/src/nupic/algorithms/TemporalMemory.cpp#L609

`activatePredictedColumn` might take in parameters `connections, excitedColumn, learn, prevActiveCells, permanenceIncrement, permanenceDecrement` and it would not take in a `self`. It would just be a function. In many ways this is what the Python `TemporalMemory` does currently, it just uses methods instead of functions... probably to support overriding.

My coding style was to put the output params at the beginning of the list.